power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/sparse_lu_solver.hpp

```diff
@@ -328,6 +405,210 @@ template <class Tensor, class RHSVector, class XVector> class SparseLUSolver {
     std::shared_ptr<IdxVector const> row_indptr_;
     std::shared_ptr<IdxVector const> col_indices_;
     std::shared_ptr<IdxVector const> diag_lu_;
+    // cache value for pivot perturbation for the factorize step
+    bool has_pivot_perturbation_{false};
+    double matrix_norm_{};
+    std::optional<std::vector<Tensor>> original_matrix_;
+    // cache value for iterative refinement for the solve step
+    std::optional<std::vector<XVector>> dx_;
+    std::optional<std::vector<RHSVector>> residual_;
+    std::optional<std::vector<RHSVector>> rhs_;
+
+    void solve_with_refinement(std::vector<Tensor> const& data,        // pre-factoirzed data, const ref
+                               BlockPermArray const& block_perm_array, // pre-calculated permutation, const ref
+                               std::vector<RHSVector> const& rhs, std::vector<XVector>& x) {
+        // initialize refinement
+        initialize_refinement(rhs, x);
+        double backward_error{10.0};
+        solve_once(data, block_perm_array, rhs_.value(), x);
+        Idx num_iter{};
+        // iterate until convergence
+        while (backward_error > epsilon_perturbation) {
```

Comment:
this criterion still misses the "still worth it" check that

$$\text{berr} < \frac{1}{2}\cdot \text{lastberr}$$

was that intentional?

Reply:
Yes. We want to the iteration go on. If it reaches maximum iteration (5), we throw Sparse Matrix Error.